SetEditor

murraystevenson opened the pull request:
This adds a first version of the SetEditor. A new panel for inspecting and (eventually) editing sets. 

![SetEditor](https://user-images.githubusercontent.com/50844517/220836831-b39b0508-9c48-46b7-a707-a6533650e997.png)

Initial functionality focuses on the inspection side of things:
- Providing a hierarchical view of set names and their members, with the hierarchy defined by splitting set names on ":". 
- Filtering of the displayed sets by name, and the ability to hide empty sets.
- Drag and drop access to set names and set members by dragging from the `Name` column or `Members` column. 
- Context menu actions for copying selected set names, and copying or selecting set members.

A few other niceties have been included:

 Drag and drop of StringVectorData onto StringPlugValueWidgets and Spreadsheet cells representing StringPlugs, converting the StringVectorData to a space separated string. This enables behaviours such as dragging of multiple set names from the SetEditor onto a SetFilter's `setExpression` plug in the Node Editor, or Spreadsheet equivalent.

Allowing right-click to select paths in a PathListingWidget, this lets you right-click on an unselected path and have the context menu act on it as it now selected.

Dragging one or more paths from the HierarchyView to the SetEditor's search filter will populate the filter with the sets those paths are a member of. It could be nice to later add a mode to the Set Editor that automatically displays the sets for the currently selected scene locations, but this also seemed like a convenient (if slightly hard to discover) way of drilling down to specific sets based on scene location(s).

This also includes the previously discussed fast path for sorting a PathModel by Name. This does introduce a bit of flicker in the PathChooserWidget as row heights and column widths can quickly update from Name only to populated with cell data. Updating the stylesheet to set a fixed item height for the PathChooserWidget like we've done previously for the HierarchyView and LightEditor seems to mitigate most of it, but let me know what you think of the file browsing experience.

johnhaddon commented on python/GafferSceneUI/SetEditor.py:
```diff
@@ -0,0 +1,379 @@
+
+		return GafferUI.NodeSetEditor._titleFormat( self, _maxNodes = 1, _reverseNodes = True, _ellipsis = False )
+
+	def __plugDirtied( self, plug ) :
+
+		if self.__plug is not None and isinstance( plug, Gaffer.PathMatcherDataPlug ) and plug == self.__plug["set"] :
+			self.__updatePathListingPath()
```
I don't think this should be necessary - see comment in `SetPath::plugDirtied()`.

murraystevenson replied:
As above, removed in 4c99596.

johnhaddon commented on python/GafferSceneUI/SetEditor.py:
```diff
@@ -0,0 +1,379 @@
+	def __selectSetMembers( self, *unused ) :
+
+		locations = self.__getSetMembers()
+		if not locations.isEmpty() :
```
I feel like it might be worth dropping this check, so that no matter what, after the action is complete, the current selection is exactly the same as the content of the set. I think confusion could ensue if you quickly use the shortcut without looking at the members column, and then assume that the current selection (the old one that was left unchanged) is in the set.

murraystevenson replied:
Sounds good to me. Changed selection behaviour in c133edd. Also gave the copy actions the same behaviour in b98dd93.

johnhaddon commented on python/GafferSceneUI/SetEditor.py:
```diff
@@ -0,0 +1,379 @@
+
+	def __dragEnter( self, widget, event ) :
+
+		self.__patternWidget.setHighlighted( True )
```
We need to check that `event.data` is non-empty StringVectorData before accepting here. Otherwise it looks like we accept _anything_, which is misleading.

murraystevenson replied:
Fixed in 7c496f0.

johnhaddon commented on python/GafferSceneUI/SetEditor.py:
```diff
@@ -0,0 +1,379 @@
+		if isinstance( event.data, IECore.StringVectorData ) and len( event.data ) > 0 :
+			if event.data[0].startswith( "/" ) :
+				# treat as paths, return all sets those paths are members of
+				setNames = self.__getSetNamesFromPaths( event.data )
```
I'm a bit wary of how open-ended the cost of calling this might be, given that we're blocking the UI till it's done. I guess it is a very specific action to perform, and unlikely to be done accidentally though. Any merit in using a BackgroundTaskDialogue so that it can be cancelled if necessary?

murraystevenson replied:
Great idea! I've added the BackgroundTaskDialogue in 4f0cbf0.

johnhaddon commented on python/GafferSceneUI/SetEditor.py:
```diff
@@ -0,0 +1,379 @@
+# _EmptySetFilterWidget
+##########################################################################
+
+class _EmptySetFilterWidget( GafferUI.Widget ) :
+
+	def __init__( self, emptySetFilter ) :
+
+		self.__enabledWidget = GafferUI.BoolWidget()
+		self.__enabledWidget.setText( "Hide Empty" )
+		self.__enabledWidget.setToolTip( "Hide sets with no members." )
+
+		GafferUI.Widget.__init__( self, self.__enabledWidget )
+
+		self.__emptySetFilter = emptySetFilter
+		self.__enabledWidget.stateChangedSignal().connect( Gaffer.WeakMethod( self.__enabledStateChanged ), scoped = False )
+
+	def __enabledStateChanged( self, widget ) :
+
+		self.__emptySetFilter.setEnabled( self.__enabledWidget.getState() )
```
Could we just use a `BasicPathFilterWidget` here instead of this custom class? If we do need a custom class, it should probably still be implemented as a subclass of `PathFilterWidget`, and update `__enabledWidget` in `_updatefromPathFilter()`.

murraystevenson replied:
Swapped over to the BasicPathFilterWidget in 007d5c1. Is this Ok to create directly, or am I better calling `PathFilterWidget.create()`?

johnhaddon replied:
Fine to create directly.

johnhaddon commented on src/GafferSceneUIModule/SetEditorBinding.cpp:
```diff
@@ -0,0 +1,275 @@
+				std::remove_if(
+					paths.begin(),
+					paths.end(),
+					boost::bind( &SetEditorSearchFilter::remove, this, ::_1 )
```
We should be able to use lambdas instead of `boost::bind()` now.

murraystevenson replied:
Swapped `boost::bind()` for lambdas in 65f28fe, much better!

johnhaddon commented on startup/gui/layouts.py:
```diff
@@ -66,10 +67,10 @@
+layouts.add( 'Standard (multi-monitor)', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Vertical, 0.963830, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.699895, ( {'tabs': (GafferUI.GraphEditor( scriptNode ), GafferSceneUI.LightEditor( scriptNode ), GafferUI.AnimationEditor( scriptNode ), GafferSceneUI.PrimitiveInspector( scriptNode )),  ...
 layouts.add( "Empty", "GafferUI.CompoundEditor( scriptNode, windowState = {'fullScreen': False, 'screen': -1, 'bound': imath.Box2f(imath.V2f(0.0479166657, 0.108269393), imath.V2f(0.782812476, 0.906223357)), 'maximized': True} )" )
-layouts.add( 'Scene', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Horizontal, 0.772298, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.256131, ( {'tabs': (GafferSceneUI.HierarchyView( scriptNode ),), 'currentTab': 0, 'tabsVisible': True}, ( GafferUI.SplitContainer.Orientation.Vertical, 0.501064, ( ( GafferUI.SplitContainer.Orientation.V ...
+layouts.add( 'Scene', "GafferUI.CompoundEditor( scriptNode, _state={ 'children' : ( GafferUI.SplitContainer.Orientation.Horizontal, 0.772298, ( ( GafferUI.SplitContainer.Orientation.Horizontal, 0.256131, ( {'tabs': (GafferSceneUI.HierarchyView( scriptNode ), GafferSceneUI.SetEditor( scriptNode )), 'currentTab': 0, 'tabsVisible': True}, ( GafferUI.SplitContainer.Orientation.Vertical, 0.501064, ( ( ...
```
Perhaps the SetEditor should be below rather than behind the HierarchyView in this layout? It seems like you'd often use them in tandem, so it's a pity to make that hard.

murraystevenson replied:
Adjusted the layout in 0e60294. It does seem an improvement over hiding the SetEditor away in a tab behind the HierarchyView.

johnhaddon commented on python/GafferUITest/PathListingWidgetTest.py:
```diff
@@ -990,6 +996,9 @@ def testModelChangingData( self ) :
 				GafferUI.PathListingWidget.StandardColumn( "Value", "dict:value" )
 			],
 		)
+		# Sort by "Value" column as sorting by defaultNameColumn doesn't
+		# trigger a data update.
+		widget._qtWidget().header().setSortIndicator( 1, QtCore.Qt.AscendingOrder )
```
I don't this test was intended to depend on sorting to trigger the initial update - I think that was just an oversight on my part. I think instead we should replace L1007-1010 with this, to be explicit about what we're querying :

```
		self.assertEqual( model.rowCount(), 0 )
		_GafferUI._pathModelWaitForPendingUpdates( GafferUI._qtAddress( model ) )
		self.assertEqual( model.rowCount(), 1 )
		self.assertEqual( model.data( model.index( 0, 1, QtCore.QModelIndex() ) ), None )
		_GafferUI._pathModelWaitForPendingUpdates( GafferUI._qtAddress( model ) )
		self.assertEqual( model.data( model.index( 0, 1, QtCore.QModelIndex() ) ), 10 )
```

murraystevenson replied:
Fixed in a1b7bd3.

johnhaddon commented on python/GafferUITest/PathListingWidgetTest.py:
```diff
@@ -950,6 +950,12 @@ def testModelFirstQueryDoesntEmitDataChanged( self ) :
+				# Sort by "Value" column as sorting by defaultNameColumn
+				# doesn't trigger a data update.
+				widget._qtWidget().header().setSortIndicator( 1, QtCore.Qt.AscendingOrder )
+				_GafferUI._pathModelWaitForPendingUpdates( GafferUI._qtAddress( model ) )
```
Why do we need another wait here? If we want to honour the intentions of the original test, shouldn't we set the sort indicator at L939 and expect that the existing waits are sufficient?

murraystevenson replied:
Fixed in 442c5f3.

johnhaddon commented on python/GafferUITest/PathListingWidgetTest.py:
```diff
@@ -1153,14 +1162,18 @@ def headerData( self, canceller = None ) :
 				]
 			)
 
+		# Sort by SleepingColumn as sorting by defaultNameColumn doesn't
+		# trigger a data update.
+		widget._qtWidget().header().setSortIndicator( 1, QtCore.Qt.AscendingOrder )
```
I think the intention in this test was that showing the window would cause Qt to query the data so it could be displayed. changing to `waitForIdle( 1000 )` on L1179 seems to be sufficient to make that happen without messing about with sorting.

murraystevenson replied:
Fixed in 5f36636, though I needed to go all the way to `waitForIdle( 20000 )` in order for the test to reliably pass on my workstation...

johnhaddon commented on python/GafferUITest/PathListingWidgetTest.py:
```diff
@@ -1277,6 +1290,53 @@ def testStretchLastColumnIsSetFromColumns( self ) :
+		# setting our widget to be sortable shouldn't result in any data changes
+		# as we're sorting by the `name` column
+
+		widget.setSortable( True )
+		_GafferUI._pathModelWaitForPendingUpdates( GafferUI._qtAddress( model ) )
+		self.assertEqual( len( changes ), 0 )
+
+		# likewise, changing the sort order of the `name` column shouldn't result
+		# in data changes
+
+		widget._qtWidget().header().setSortIndicator( 0, QtCore.Qt.AscendingOrder )
+		_GafferUI._pathModelWaitForPendingUpdates( GafferUI._qtAddress( model ) )
+		self.assertEqual( len( changes ), 0 )
```
Would be nice to assert that the model has actually been sorted into the expected order here.

murraystevenson replied:
Added testing of expected sort orders (and swapped the second sort to Descending in order to actually see a change 🙃 ) in 2de42a1.

johnhaddon commented on python/GafferUI/PathListingWidget.py:
```diff
@@ -620,7 +620,7 @@ def __buttonPress( self, widget, event ) :
 		# Otherwise, perform expansion and selection as appropriate.
 
-		if event.buttons != event.Buttons.Left :
+		if event.buttons not in ( event.Buttons.Left, event.Buttons.Right ) :
```
It feels a bit odd that we can accept the right-click here but that the context menu signal still gets emitted. Is that consistent on all platforms (I think context menu signal has some per-platform differences)?

murraystevenson replied:
Ugh. This behaves reasonably on Linux and macOS, but I gave it a try on Windows and as you suspected the combination is sub-optimal. Right-clicking within a range of already selected items results in the selection updating to just the item clicked. As this behaviour isn't crucial to this PR maybe I'll drop this for now and we can revisit later rather than holding the rest up.

johnhaddon reviewed (changes requested):
Thanks Murray, this is great! I've made a few comments inline, but hopefully none of them are problematic...

murraystevenson commented on src/GafferScene/SetPath.cpp:
```diff
@@ -0,0 +1,323 @@
+
+void SetPath::plugDirtied( Gaffer::Plug *plug )
+{
+	if( plug == m_scene->setNamesPlug() )
```
Addressed in 9b7e4f0 and the manual update removed in 4c99596.

murraystevenson commented on include/GafferScene/SetPath.h:
```diff
@@ -0,0 +1,113 @@
+
+IE_CORE_FORWARDDECLARE( ScenePlug )
+
+class GAFFERSCENE_API SetPath : public Gaffer::Path
```
Until such a moment as we find a general use-case for SetPath to justify its wider existence, I've moved it over to GafferSceneUI as part of SetEditorBinding.

Two fixups here, one to revert the initial addition of SetPath to GafferScene 3ee6763, and another to add it to GafferSceneUI d84b2a3. Once we're all good with this reshuffle, I'll drop the `GafferScene : Add SetPath` related commits as part of the squash.

Do you think this goes far enough or should we also have Python access to SetPath housed within `GafferSceneUI.Private`?

johnhaddon replied:
> Do you think this goes far enough or should we also have Python access to SetPath housed within GafferSceneUI.Private?

I don't think it goes far enough, because it's currently accessible as `GafferSceneUI.SetPath`, which is public. The precedent from LightEditor and the other bits of SetEditor would be for it to be bound as `GafferSceneUI._SetEditorSetPath`, which is reasonably private. I think that'd do for now? Alternatively, we could make `GafferSceneUI._SetEditor.SetPath` and move the other set editor C++ components into that new namespace.

murraystevenson commented on python/GafferSceneUI/SetEditor.py:
```diff
@@ -0,0 +1,379 @@
+	def __updatePathListingPath( self ) :
+
+		if self.__plug is not None :
+			# We take a static copy of our current context for use in the SetPath - this prevents the
+			# PathListing from updating automatically when the original context changes, and allows us to take
+			# control of updates ourselves in _updateFromContext(), using LazyMethod to defer the calls to this
+			# function until we are visible and playback has stopped.
```
I've added a todo to revisit this in aa00db2.

johnhaddon reviewed (commented):
Thanks for the updates Murray! I think we're good to go, minus the privatisation of the SetPath binding. Could you do that at the same time as squashing down please?

murraystevenson commented:
Ok, we should be good to merge. Squashed, dropped, privatised and updated to use `#pragma once` for good measure. I ended up taking the alternate suggestion for privatising the SetPath binding as PathClass didn't seem to provide a method of overriding the class name (or it's quite possible I missed something obvious).

johnhaddon commented:
I've rebased to fix a conflict in Changes.md, and merged. Thanks Murray!
